File: n-queens.py
from typing import List
class Solution:
    def solveQueens(self, n: int) -> List[List[str]]:
        col = set()
        posDiag = set()  # (r+c)
        negDiag = set()  # (r-c)

        res = []
        #We create a board of n + n :
        board = [["."] * n for i in range(n)]

        def backtrack(r):  # We go row per row
            if r == n:  # true when we are completed
                copy = ["".join(row) for row in board]
                res.append(copy)
                return
            for c in range(n):  # For each position(column) in the row
                # We look if current column, positive or negative diagonals have been already used
                #We use continue as we are not allowed to use this column in the True case.
                if c in col or (r+c) in posDiag or (r-c) in negDiag:
                    continue

                col.add(c)
                posDiag.add(r+c)
                negDiag.add(r-c)
                board[r][c] = "Q"
                backtrack(r + 1)  # Next row

                #We do a clean up in case we go back to search for all possible solutions!
                col.remove(c)
                posDiag.remove(r+c)
                negDiag.remove(r-c)
                board[r][c] = "."
                #We do not have a return in this loop so once it is done means the function finished
        backtrack(0)
        return res

solver = Solution()
result = solver.solveQueens(4)
print(result)
for config in result:
    for row in config:
        print(row)
    print("\n")

File: nQueens.go
package Main

import (
	"fmt"
	"strings"
)

func main() {
	solver := Solution{}
	result := solver.solveQueens(4)
	fmt.Println(result)
	for _, config := range result {
		for _, row := range config {
			fmt.Println(row)
		}
		fmt.Println()
	}
}

type Solution struct {
	col     map[int]bool
	posDiag map[int]bool
	negDiag map[int]bool
	res     [][]string
	board   [][]string
}

func (s *Solution) solveQueens(n int) [][]string {
	s.col = make(map[int]bool)
	s.posDiag = make(map[int]bool)
	s.negDiag = make(map[int]bool)
	s.board = make([][]string, n)
	for i := 0; i < n; i++ {
		s.board[i] = make([]string, n)
		for j := 0; j < n; j++ {
			s.board[i][j] = "."
		}
	}
	s.backtrack(0, n)
	return s.res
}

func (s *Solution) backtrack(r int, n int) {
	if r == n {
		copy := make([]string, len(s.board))
		for i, row := range s.board {
			copy[i] = strings.Join(row, "")
		}
		s.res = append(s.res, copy)
		return
	}

	for c := 0; c < n; c++ {
		if s.col[c] || s.posDiag[r+c] || s.negDiag[r-c] {
			continue
		}
		s.col[c] = true
		s.posDiag[r+c] = true
		s.negDiag[r-c] = true
		s.board[r][c] = "Q"
		s.backtrack(r+1, n)

		s.col[c] = false
		s.posDiag[r+c] = false
		s.negDiag[r-c] = false
		s.board[r][c] = "."
	}
}
